### PR TITLE
Request lifecycle doc edits

### PR DESCRIPTION
**Pull request by Meschreiber**
This PR makes some edits to https://github.com/apollographql/router/pull/3391.

Fixes #issue_number

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: .changesets/docs_geal_document_request_lifecycle.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 ### Document the request lifecycle ([PR #3391](https://github.com/apollographql/router/pull/3391))
 
-This adds an in depth documentation of the entire request lifecycle, which services exist in the router, the request and response types they use, and where plugins can attach themselves
+This adds in-depth documentation of:
```
Comment (Meschreiber) [inline .changesets/docs_geal_document_request_lifecycle.md:3]: If a sentences includes a list of 3+ items, it's worth considering if it would be clearly as a list.

File: docs/source/customizations/overview.mdx [hunk 2]
```diff
@@ -117,4 +95,31 @@ class client,subgraph1,subgraph2 secondary;
...
+
+The supergraph service works on a `SupergraphRequest` containing HTTP headers and a GraphQL request object. The `SupergraphResponse` contains headers and a stream of GraphQL responses. That stream only contains one element for most queries—it can contain more if the query uses the `@defer` directive or subscriptions.
+
+The supergraph service calls into the query planner, which returns a query plan. The supergraph service then calls the execution service.
+
+### Execution service
+
+The execution service is tasked with executing the query plan. The `ExecutionRequest` contains the `SupergraphRequest` and the query plan. The `ExecutionResponse` has the same content as the `SupergraphResponse`.
+
+For each fetch node of the query plan, the execution service creates a subgraph request and then calls the subgraph plugins and service. Once the execution service has received all subgraph responses, it formats the GraphQL responses—removing unneeded data and propagating nulls—before sending it back to the supergraph plugin.
```
Comment (Meschreiber) [inline docs/source/customizations/overview.mdx:112]: Does the execution service also compose the subgraph responses into one response? Or does it send them separately?

```suggestion
For each fetch node of the query plan, the execution service creates a subgraph request and then calls the subgraph plugins and service. Once the execution service has received all subgraph responses, it formats the GraphQL responses—removing unneeded data and propagating nulls—before sending them back to the supergraph plugin.
```

File: docs/source/customizations/overview.mdx [hunk 3]
```diff
@@ -10,7 +10,7 @@ You can create **customizations** for the Apollo Router to add functionality tha
 The Apollo Router supports the following customization types:
 
 - [**Rhai scripts**](./rhai/)
-    - The [Rhai scripting language](https://rhai.rs/book/) enables you to add functionality directly to your stock router binary by hooking into different phases of the router's request lifecycle.
+    - The [Rhai scripting language](https://rhai.rs/book/) lets you add functionality directly to your stock router binary by hooking into different phases of the router's request lifecycle.
 - [**External co-processing**](./coprocessor/) ([Enterprise feature](../enterprise-features/))
     - If your organization has a [GraphOS Enterprise plan](https://www.apollographql.com/pricing/), you can write custom request-handling code in any language. This code can run in the same container as your router or separately.
     - The router calls your custom code via HTTP, passing it the details of each incoming client request.
```
Comment (Meschreiber) [inline docs/source/customizations/overview.mdx:16]: Why do we recommend using Rhai scripts if possible?

File: docs/source/customizations/overview.mdx [hunk 4]
```diff
@@ -23,98 +23,182 @@ The Apollo Router supports the following customization types:
...
----
+1. The router receives a client request at an HTTP server. 
+2. The HTTP server transforms the HTTP request into a `RouterRequest` containing HTTP headers and the request body as a stream of byte arrays.
+3. The router service receives the `RouterRequest`. It handles Automatic Persisted Queries (APQ), parses the GraphQL request from JSON and calls the supergraph service with the resulting `SupergraphRequest`.
+4. The supergraph service calls the query planner with the GraphQL query from the `SupergraphRequest`.
+5. The query planner returns a query plan for most efficiently executing the query.
+6. The supergraph service calls the execution service with an `ExecutionRequest`, made up of `SupergraphRequest` and the query plan.
+7. For each fetch node of the query plan, the execution service creates a `SubgraphRequest` and then calls the respective subgraph plugins and service.
+8. Each subgraph has its own subgraph service, and each service can have its own subgraph plugin configuration. The subgraph service transforms the `SubgraphRequest` into an HTTP request to that subgraph. The `SubgraphRequest` contains:
+    - the (read-only) `SupergraphRequest`
```
Comment (Meschreiber) [inline docs/source/customizations/overview.mdx:147]: Why is this included?

**Comment by Meschreiber** [comment]
@Geal , thank you so much for this overview! It helped me understanding the inner workings of the router much better. I took the Mermaid diagram you made and turned it into a few separate diagrams for digestibility.

I left a few questions, which you're welcome to answer here, or when we chat.

My bigger question is how users can concretely use this knowledge to create customizations, whether Rhai scripts or coprocessors. Also, while I hope my restructuring makes it easier to understand the lifecycle, I'm not sure if this structure (rather than content being grouped by service and what each service does) helps the goal of making customization easier, so would be curious for your take on that.

File: docs/source/customizations/overview.mdx (hunk 3 above)
Comment (Geal) [reply docs/source/customizations/overview.mdx:16]: it's the easiest customization to deploy: users write a rhai file and load it into the router. Coprocessors need an external service that exposes the API expected by the router, and native plugins requires rebuilding the router

File: docs/source/customizations/overview.mdx (hunk 4 above)
Comment (Geal) [reply docs/source/customizations/overview.mdx:147]: it's remaining from the initial design, and something we want to remove (https://github.com/apollographql/router/issues/3400). Unfortunately it is part of the public API, so removing it right now would be a breaking change

File: docs/source/customizations/overview.mdx [hunk 5]
```diff
@@ -23,98 +23,182 @@ The Apollo Router supports the following customization types:
...
 ```
 
----
+1. The router receives a client request at an HTTP server. 
+2. The HTTP server transforms the HTTP request into a `RouterRequest` containing HTTP headers and the request body as a stream of byte arrays.
+3. The router service receives the `RouterRequest`. It handles Automatic Persisted Queries (APQ), parses the GraphQL request from JSON and calls the supergraph service with the resulting `SupergraphRequest`.
+4. The supergraph service calls the query planner with the GraphQL query from the `SupergraphRequest`.
+5. The query planner returns a query plan for most efficiently executing the query.
+6. The supergraph service calls the execution service with an `ExecutionRequest`, made up of `SupergraphRequest` and the query plan.
+7. For each fetch node of the query plan, the execution service creates a `SubgraphRequest` and then calls the respective subgraph plugins and service.
```
Comment (Geal) [inline docs/source/customizations/overview.mdx:145]: the issue in separating the request and response sides here is that in the execution service, subgraph requests can happen both in parallel and in sequence, because one subgraph response's data can be used in another subgraph request

**Comment by Meschreiber** [comment]
@Geal , thanks for the helpful chat today! I added in some of the information we discussed. I'm actually now not sure if we need to include further information that could be gleaned from a stack trace. I think it may be enough to point out that there are some "nuances" to the request/response cycle and that any customizations you create will be affected by those nuances.

Whenever you get a chance, take a look at this most recent version and lmk what you think. (You may not need to go through the trouble of sending me a trace :) ).

**Comment by Geal** [comment]
Screenshot from 2023-08-09 15-52-34

Would something like that work?

**Review by Geal — APPROVED** [review]

**Review by garypen — APPROVED** [review]
I liked Geoffroy's documentation and I like your improvements to it. Nice work both of you!

**Comment by Meschreiber** [comment]
> Would something like that work?
I think we may not need it. Let me know if you feel otherwise.

Otherwise, I think this is ready to merge except for the Mermaid upgrade required to get the diagram just right. Let me see if I can get to that today!

**Comment by Meschreiber** [comment]
Just realized Mermaid package updates will happen in the docs repo so merging this.

**Comment by Geal** [comment]
awesome, thanks for your help!

**Comment by Meschreiber** [comment]
Looks like there's some breaking changes for [updating Mermaid](https://github.com/apollographql/docs/pull/519) that DXE will address later. Though not ideal, the current diagram is OK IMO.
